Implementation of DateTime2 + accompanying GetSqlDateTime2 method (#846)

bjornbouetsmith opened the pull request:
Initial implementation - let me know what you think so far.
I have tried to do everything I could think of and have tried to follow the "style"
I don't know why VS insists on changing the .csproj for SqlCommand etc - I have done nothing, but will try to revert the changes out side of VS

And I need some help understanding why VS can build my stuff when I use the ADP class, but I cannot build it outside of VS - possibly some "tweaks" needed - which is why I am throwing exceptions directly and not going through the ADP class like the other classes do.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=853) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=853) before we can accept your contribution.<br/><br/>:x: bjornbouetsmith [sign now](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=853)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/SqlClient?pullRequest=853) it.</sub>

JRahnama commented on src/Microsoft.Data.SqlClient/netfx/src/Microsoft.Data.SqlClient.csproj:
```diff
@@ -355,18 +358,26 @@
-    <Compile Include="Microsoft\Data\SqlClient\SqlCommandBuilder.cs" />
+    <Compile Include="Microsoft\Data\SqlClient\SqlCommand.cs">
+      <SubType>Component</SubType>
+    </Compile>
```
On initial review, remove these `SubType` VS adds them by default. It happens in several places. Check other similar tags in the same file.

bjornbouetsmith replied:
Yeah - I will do that last, when all other review changes has been done, since VS will probably add them in again - annoyingly enough.

JRahnama commented on src/Microsoft.Data.SqlClient/netfx/src/Microsoft.Data.SqlClient.csproj:
```diff
@@ -355,18 +358,26 @@
     <Compile Include="Microsoft\Data\SqlClient\SqlCredential.cs" />
-    <Compile Include="Microsoft\Data\SqlClient\SqlDataAdapter.cs" />
+    <Compile Include="Microsoft\Data\SqlClient\SqlDataAdapter.cs">
+      <SubType>Component</SubType>
```
Same here.

JRahnama commented on src/Microsoft.Data.SqlClient/netfx/src/Microsoft.Data.SqlClient.csproj:
```diff
@@ -432,7 +443,9 @@
     <Compile Include="Microsoft\Data\ProviderBase\DbConnectionFactory.cs" />
-    <Compile Include="Microsoft\Data\ProviderBase\SqlConnectionHelper.cs" />
+    <Compile Include="Microsoft\Data\ProviderBase\SqlConnectionHelper.cs">
+      <SubType>Component</SubType>
```
same here

JRahnama commented on src/Microsoft.Data.SqlClient/netfx/src/Microsoft.Data.SqlClient.csproj:
```diff
@@ -355,18 +358,26 @@
     <Compile Include="Microsoft\Data\SqlClient\SqlCommandSet.cs" />
-    <Compile Include="Microsoft\Data\SqlClient\SqlConnection.cs" />
+    <Compile Include="Microsoft\Data\SqlClient\SqlConnection.cs">
+      <SubType>Component</SubType>
```
same here.

JRahnama commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,138 @@
+using System.Xml;
+using System.Xml.Schema;
+using System.Xml.Serialization;
+//using Microsoft.Data.Common;
```
Remove on used libraries.

bjornbouetsmith replied:
I would never leave it in, but VS can build with that using and using the ADP class - when I try to build outside of VS using the `msbuild /T:Release` command it fails with the error that it cannot find that namespace - so when I have had help figuring out why I cannot build - I will include it again and also the ADP class

Wraith2 replied:
my build commands are:
```
msbuild /t:clean
msbuild /p:Configuration=Release
msbuild /t:BuildTestsNetCore /p:TargetNetCoreVersion=netcoreapp3.1 /p:Configuration=Release
```
without some of those args it can get a bit confused about what it's building and why.

bjornbouetsmith replied:
No dice - I still get:

````
"D:\git\SqlClient\build.proj" (default target) (1) ->
"D:\git\SqlClient\src\Microsoft.Data.SqlClient\netcore\src\Microsoft.Data.SqlClient.csproj" (default target) (4:33) ->
"D:\git\SqlClient\src\Microsoft.Data.SqlClient\netcore\src\Microsoft.Data.SqlClient.csproj" (Build target) (4:36) ->
  D:\git\SqlClient\src\Microsoft.Data.SqlClient\src\Microsoft\Data\SqlTypes\SqlDateTime2.cs(7,22): error CS0234: The type or namespace name 'Common' does not exist in the namespace 'Microsoft.Data' (are you missing an assembly reference?) [D:\git\SqlClient\src\Microsoft.Data.SqlClient\netcore\src\Microsoft.Data.SqlClient.csproj]
````

If you have a local copy, can you try to mimic what I have done, i.e. in a class that lives in the `src\Microsoft.Data.SqlClient\src` common area include 
using Microsoft.Data.Common;
and reference the ADP class - just to see if you can reproduce it.

bjornbouetsmith replied:
I have even tried now mimicking the azure build command line:
```
msbuild build.proj /nologo /nr:false /t:BuildAllConfigurations /p:GenerateNuget=false /p:configuration="Release"
```
And that does not work either - it seems like visual studio does some "magic" with the common files linked in.

I will investigate further by adding diagnostics logging to msbuild to see why it seems like its not picking up the namespace/ADP class

bjornbouetsmith replied:
I think this might the the reason why its not working for me:

in the Microsoft.Data.SqlClient.csproj for netcore:

```
<ItemGroup Condition="'$(OSGroup)' != 'AnyOS'">
```

Where the file for the ADP class is defined under - so I guess I would need to move SqlDateTime2.cs up in the same ItemGroup.

JRahnama commented on src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlDateTime2Test.cs:
```diff
@@ -0,0 +1,45 @@
+            SqlDateTime2 sqlDateTime2 = (SqlDateTime2)value;
+
+            Assert.Equal(expected, sqlDateTime2);
+
```
Remove extra line here.

JRahnama commented on src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlDateTime2Test.cs:
```diff
@@ -0,0 +1,45 @@
+            DateTime converted = (DateTime)expected;
+
+            Assert.Equal(value, converted);
+
```
Extra line.

JRahnama commented on src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DateTime2Test/DateTime2Test.cs:
```diff
@@ -0,0 +1,90 @@
+            
+            Assert.Equal(TestDate, value.Value);
+        }
+
```
double line is not needed. Remove the extra space here please.

JRahnama commented:
Some of the test failures are not due to your changes. You can address the ones are related to your changes (if there is any) and push the changes again.

bjornbouetsmith commented:
> Some of the test failures are not due to your changes. You can address the ones are related to your changes (if there is any) and push the changes again.

Yeah I looked at one of the test failues and that has definately not anything to do with my additions - but will go through them all just to be sure when they are all finished.

Wraith2 commented on src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBuffer.cs:
```diff
@@ -630,6 +630,41 @@ internal SqlDateTime SqlDateTime
 
+        internal SqlDateTime2 SqlDateTime2
+        {
+
```
nit: remove empty line

bjornbouetsmith replied:
sorry, missed that one

Wraith2 commented on src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlBuffer.cs:
```diff
@@ -602,6 +602,41 @@ internal SqlDateTime SqlDateTime
 
+        internal SqlDateTime2 SqlDateTime2
+        {
+
```
nit: remove empty line

bjornbouetsmith replied:
sorry, missed that one

Wraith2 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,138 @@
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlDateTime2.xml' path='docs/members[@name="SqlDateTime2"]/ctor2/*' />
+        public SqlDateTime2(long ticks)
+        {
+            if (ticks < m_minTicks && ticks > m_maxTicks)
```
Shouldn't this be `||`?

bjornbouetsmith replied:
Of course :-)

Wraith2 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,138 @@
+        public static readonly SqlDateTime2 Null = new SqlDateTime2(true);
+
+        /// <inheritdoc />
+        public bool Equals(SqlDateTime2 other)
```
null check other or ?. access

Wraith2 commented on src/Microsoft.Data.SqlClient/tests/FunctionalTests/Microsoft.Data.SqlClient.Tests.csproj:
```diff
@@ -60,6 +61,10 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkVersion)" />
     <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="$(SystemDiagnosticsDiagnosticSourceVersion)" />
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonVersion)" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
```
Not sure this should be needed if it wasn't there already.

bjornbouetsmith replied:
I could not run any tests within visual studio without it.

bjornbouetsmith replied:
But I can remove it just before I squash the final push if you think?

karinazhou replied:
Please remove this PackageReference before the final merge. We have 
`<PackageReference Include="xunit.runner.visualstudio" Version="$(XunitVersion)" />`
in src\Microsoft.Data.SqlClient\tests\tools\Microsoft.DotNet.XUnitExtensions\Microsoft.DotNet.XUnitExtensions.csproj which will be picked up by msbuild.

bjornbouetsmith replied:
ok, it just makes it impossible to run any tests from within visual studio, which is a shame.
But will do.

Wraith2 replied:
Odd. I thought the tests ran inside visual studio before. If not then can you file an issue for it? I run the tests from the commandline because that's how I'm used to working with the project from back when it was part of corefx.

Wraith2 reviewed (commented):
Some comments added inline but overall looks good to me.

bjornbouetsmith commented:
Checked the tests and it is all relating to encryption in the ManualTests - so has nothing to do with the changes I have made.

bjornbouetsmith commented:
Changed the internals of SqlDateTime2 to be inline with the https://github.com/dotnet/SqlClient/blob/master/coding-style.md - and since its a new file I changed some of the fieldnames.

bjornbouetsmith commented:
squashed everything into one commit

bjornbouetsmith commented:
Any chance to get this reviewed and merged so I don't have to handle any more conflicts?

cheenamalhotra commented:
Hi @bjornbouetsmith 

We will be reviewing this PR soon, currently we're occupied with a bundle of support issues that have much higher priority.
We will definitely try to come up some update by end of this month.

Thanks for your patience!

cheenamalhotra reviewed (commented):
Partially reviewed.

We need much more tests I think. e.g.  Async Readers, and anything else that the new type is going to add support for.

Also as documented here: https://docs.microsoft.com/en-us/sql/connect/ado-net/sql-server-data-type-mappings?view=sql-server-ver15, `DateTime2` corresponds to `DateTime` for .NET Framework type, and is supported with for `DbType` accessor with `SqlDataReader.GetDateTime` API.

Do we have a solution without causing a breaking change that we can introduce this type, and also ensure `DateTime2` continues to work for existing apps with `DateTime` but does correspond to `DateTime2` by default in all cases to avoid confusions to new developers?

bjornbouetsmith commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,136 @@
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlDateTime2.xml' path='docs/members[@name="SqlDateTime2"]/ctor2/*' />
+        public SqlDateTime2(long ticks)
+        {
+            if (ticks < _minTicks || ticks > _maxTicks)
```
The point of this change was to improve performance.

Having DateTime.MinValue.Ticks & DateTime.MaxValue.Ticks in two constants is much faster than having to access DateTime.MinValue.Ticks. If you insist on referencing the DateTime.MinValue/MaxValue - a compromise could be to introduce two static fields instead of two constants, i.e. 

```c#
private static readonly long _minTicks = DateTime.MinValue.Tics;
private static readonly long _maxTicks = DateTime.MaxValue.Tics;
```

But I would prefer the two constants, since they would definetely be inlined when compiled - and I am not sure about the DateTime.Min/MaxValue.

Wraith2 replied:
I did some investigation on this a while back when I was looking over the change and it is right that there's a significant performance difference from using the constant. At some point i was going to look at the existing datetime code to see if we could benefit from using it there.
 
Thinking about that though the value that is being fetched is a constant to the jit should see that and inline it on it's own. I may ask about why it isn't in runtime.

bjornbouetsmith replied:
The Ticks property on the DateTime struct is not a constant, its calculated. DateTime.MinValue is a static value, so that is not a constant either. So the compiler cannot really inline anything. Furthermore I think when accessing the Ticks property - you have to copy the entire structure in memory because its not readonly.

cheenamalhotra replied:
Can we do "MinTicks" and "MaxTicks" instead as they're constants?

https://github.com/dotnet/runtime/blob/e6a5339b9231976eccd29abb1c58298e353e4e87/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L79-L80

bjornbouetsmith commented:
I am not sure why you need many more tests.

The SqlDateTime2 struct is only being exposed in one single method: GetSqlDateTime2 - and does not interface with any of the existing methods.

So its a non breaking change - available only in SqlDataReader - if support had been added in DbDataReader and all over the place, it would be a different change.

But this change was a needle point expansion of SqlDataReader only.

I had no plans of introducing usage of SqlDateTime2 anywhere else, but if anyone else wants to do that, then yes more tests are needed - but I think the tests I made for GetSqlDateTime2 is more than sufficient and should handle the possible scenarios.

If not, I am happy to write more tests.

My change was made, so I could stop using GetDateTime and use GetSqlDateTime2 instead - similar to GetSqlDateTime - which unfortunately does not support the DATETIME2 type.

bjornbouetsmith commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,136 @@
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SqlDateTime2 : INullable, IComparable<SqlDateTime2>, IEquatable<SqlDateTime2>, IComparable, IXmlSerializable
+    {
+        private const long _minTicks = 0; // DateTime.MinValue.Ticks
+        private const long _maxTicks = 3155378975999999999; // DateTime.MaxValue.Ticks
+
```
See my comment above

slang25 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs:
```diff
@@ -0,0 +1,134 @@
+        private const long _maxTicks = 3155378975999999999; // DateTime.MaxValue.Ticks
+
+        private bool _notNull;    // false if null - has to be _notNull and not _isNull - otherwise default(SqlDateTime2) will return a SqlDateTime2 with _isNull=false and that would make XmlSerialization fail
+        private long _ticks; // Ticks representation similar to DateTime.Tics
```
```suggestion
        private long _ticks; // Ticks representation similar to DateTime.Ticks
```

bjornbouetsmith commented:
Hi,

I am going to abandon this.

It has now been more than 3 months to get a tiny patch merged into this repository and if you are not interested enough to allocate the very few resources to get this pull request merged then so be it.

I will not be spending more time on updating the pull request as other merge into the repository and I have to handle conflicts.

So what is there now is it.

Take it or leave it - I will not be working to move this further along your apparenly very long road into the repository.

Thank you for your time so far.

cheenamalhotra commented:
Hi @bjornbouetsmith 

Apologies I missed your earlier message.

I'd just like to state, in order to make a public API introduction like this one, we need a stronger case and not enough people have shown interest in this PR so far, neither in issue https://github.com/dotnet/SqlClient/issues/846. We don't want to discourage customers, but this is the reason why this feature loses priority over other customer issues. If there had been a lot of interest from community, this PR would have got desired attention as it would have bubbled up priority score.

Thanks for your suggestion, if we get more interest on issue #846 , we will consider re-opening this again!

Wraith2 commented:
How much external attention is required to get a new function added? There are a few things I'd like to be able to add or support adding (like these functions) but the number of people involved in this repository for anything other than bug reporting is extremely low so getting more than one of two upvotes is very unlikely. 

Just how big a crowd do I need to hire to get a new feature added?

cheenamalhotra commented:
@Wraith2 it depends on various factors.

This PR introduces a new **SqlDbType** and also a new API to append the core driver usability on `DataReader` where no one else has yet raised a concern and alternate solutions already exist, so it does require strong support in order to move it forward. Our customer base also includes SSMS, Azure Data Factory, etc, and we have not heard any such need, so it's very unlikely there's a requirement for a new API.

Also if you see in issue https://github.com/dotnet/SqlClient/issues/1009#issuecomment-808694631 discussions, the trends do not favor adding new methods/APIs for each and every type that don't add considerable benefit and are proceeding to generic behaviors. So unless and until there's a huge roadblock impacting a major customer base, we'd also know from within, and only then such major decisions can be made. cc @David-Engel @saurabh500

Wraith2 commented:
I think that many consumers assume that the api's for dealing with extra types won't be added because it's been so long since they were made available in sql server and they're still not in this driver. There are very few people interested in ADO advancements because they've been ignored for so long so for any api proposal outside the MS teams associated with the product there are very few people who are going to see it. Realistically how many frequent users are there on this particular product? half a dozen at most and I'd struggle to name many beyond Erik and myself.

I intend to implement the additional api we recently investigated with named parameters in large parameter sets. Do I need to get signoff on that before starting? And if so how do i go about doing that and how much public support do I ned to demonstrate that it is worth adding? The report we investigated was the only one that I've seen on the topic so it's low use but it's a clear improvement to the driver and allows big data users to be more efficient.

To return to the datetime2 api in this PR it isn't feasible to add it to the GetFieldValue members because it returns a specialized structure due to the range differences. If it were possible it would have been suggested. I see this approach as the only way to add it and it's disappointing that we can't get it included.

cheenamalhotra commented:
>  I see this approach as the only way to add it and it's disappointing that we can't get it included.

I'd like to re-iterate, I didn't mean to discard the PR or API neither did I say it cannot be included.

As I said before, this wasn't in our priority list due to not enough demand and already available workarounds, but it has been reviewed for sanitization. We have lots of priority customer issues that eventually push this behind to be polished and considered for a release milestone.
